###This notebook first simulates and then solves a game I play with my young daughter often. In this game you roll a die, if you roll a die with a color you take a fruit of that color from its tree. There are 4 types of these fruits. One other side of the die is the raven. If you roll the raven 5 times before getting all the fruits from the trees you lose. The last die is a basket where you choose which color you would like to would like to take from the tree. And this choice is what is being investigated in terms of win percentages. 

In [3]:
import numpy as np

from first_orchard_solver.gameplay.gamesims import run_batches
from first_orchard_solver.gameplay.gamesolver import win_perc, win_perc_comp
from first_orchard_solver.gameplay.gamelogic import GameState

In [19]:
game_state = GameState()
batches = run_batches(game_state, 100, 10000, strat=["fewest", "most", "random"])


###run_batches simulates playing many batches of games using different strategies. The different strategies pertain to when you have a choice of moving a fruit of your choice from the basket, you can choose one fruit at random, from one of the fruit types that have the most left, or from one of the fruit type that have the least left. run_batches lets the user run x number of batches of x number of games. In the above code I set it for 10000 batches of 100 games for a total of 1000000 games. And this is for each strategy for a grand total of 3000000 games played between three different strategies. Fuller testing is in test_game_solver.py and logs for this testing are in test_validation.log for most items in this notebook, but with less verbal explanation.

In [20]:
fewest_mean = np.mean(batches.fewest_strat_runs)
print(f"Using the strategy of picking from the fruit with the least remaining on a " 
      f"wild roll \nhas an estimated average win rate of: {fewest_mean}%")

Using the strategy of picking from the fruit with the least remaining on a wild roll 
has an estimated average win rate of: 55.5369%


In [22]:
least_mdp = win_perc((4,4,4,4), 5, "fewest")
least_mdp[0]*100

55.50000000000001

###win_perc mathematically solves the odds of winning from any game position. The solved code is usually within rounding error of the Monte Carlo simulation above. 

In [23]:
most_mean = np.mean(batches.most_strat_runs)
print(f"Using the strategy of picking from the fruit with the most remaining on a " 
      f"wild roll \nhas an estimated average win rate of: {most_mean}%")

Using the strategy of picking from the fruit with the most remaining on a wild roll 
has an estimated average win rate of: 63.069%


In [24]:
most_mdp = win_perc((4,4,4,4), 5, "most")
most_mdp[0]*100

63.2

###This is also usually within rounding error of the Monte Carlo simulation above. This mathematically proves an approximate (due to rounding) 63.2 percent chance of winning when using a consistent strategy of picking from the fruit with the most left when a wild is rolled.

In [25]:
random_mean = np.mean(batches.random_strat_runs)
print(f"Using the strategy of picking a random fruit on a " 
      f"wild roll \nhas an estimated average win rate of:  {random_mean}%")

Using the strategy of picking a random fruit on a wild roll 
has an estimated average win rate of:  59.6962%


In [26]:
rand_mdp = win_perc((4,4,4,4), 5, "random")
rand_mdp[0]*100


59.699999999999996

###For obvious reasons this will on average have more varaiation, in fact it is not strictly a solver as I used random choice in the code. However, this will still always be very close to the Monte Carlo simulation, suggesting correct calculations. 

###Now we are ready to actually compare states. The function below considers a scenario where you might be faced with a choice of which fruit to choose. Let's say you roll a wild on your second roll. Such that you have a choice to have either two fruits with 4 remaining and two fruits with 3 remaining or 3 fruits with 4 remaining and 2 of one fruit remaining. What are your odds of winning depending on which strategy you choose. Well most of the work has already been done so I just created a function that directly compares two states and it appears below. A playable version is available in main.py that will show you odds as you play.

In [ ]:
game_state1 = GameState()
game_state1.fruit_inventory.fruit_inventory = {3:4, 4:4, 5:3, 6:3}
game_state1.raven_track.spaces = 5
game_state2 = GameState()
game_state2.fruit_inventory.fruit_inventory = {3:4, 4:2, 5:4, 6:4}
game_state2.raven_track.spaces = 5
odds = win_perc_comp(game_state1, game_state2)
print(f"In this scenario, you improve your odds by {odds[0]} by choosing the fruit with" 
      f" the most remaining. Odds of the f")

(1.0999999999999943, 69.1, 70.19999999999999)

###In the scenario above we find that choosing the largest in this scenario assuming continued perfect play you will improve your odds of winning by 1.10% to 70.20% over the subpar choice which gives us a 69.10% of winning. The next step I am going to work on is to program a game with the outputs to see all of this as you play.

In [ ]:
win_perc((4,4,4,4,), 6, "most")

###This last cell shows what the percent chance of winning with various strategies altering the initial game conditions by adding one more space on the raven track. For a kids game I think 75% win rate (assuming )